docs/rfcs/20220408_partition_hibernation.md

```diff
@@ -0,0 +1,277 @@
+reporting the highest offset back to the controller so that it can
+reliably store the frozen offset state of the topic when storing
+the controller message that finalizes the hibernation.
+
+Once the topic is hibernated, all its partitions will report a
+fictional leader, so that clients know where to send metadata requests,
+and so that monitoring systems do not incorrectly flag the partition
+as being in a problematic state.
+
+On startup, a hibernated partition's storage resources are not
+recovered or replayed.  Hibernated partitions are not subject to
+continuous retention policies, BUT the controller can notice if
+a partition has been hibernated for longer than its retention.ms
+limit, and wake it up for long enough to do housekeeping.
+
+We should still be able to serve metadata requests for idle topics,
+because otherwise any monitoring system will end up re-awakening
+the topics whenever querying them.
+
+## How do we decide when to hibernate?
+
+We don't persist an `atime` type stat for partitions, and rightly so,
+because otherwise we'd be writing metadata every time someone did
+a read.
+
+Detecting an idle partition at runtime therefore relies on a partition
+staying up with stable leadership long enough to reach the idle timeout.
+That's fine if the timeout is a few hours long (if we're changing leaders
+more often than that then we've got bigger problems) but more problematic
+if we have a long idle timeout (perhaps a day or more) with a reasonable
+chance of leadership changes or node restarts before we reach it.
+
+If we wanted to support very long idle timeouts better, we could
+do so by adding something like a kvstore record of the access time
+that is updated at most every N minutes, and when a leader is elected
+it primes its access time from that if it is the same node that was
+also the previous leader.
+
+## What resources are we saving by hibernating?
+
+- Avoid creating new files on log roll during startup
+- Avoid replaying segments during startup
+- No memory overhead from storing segment indices, list of segments, etc.
+- No raft heartbeats for hibernated topics
+- No data for hibernated topics in node health telemetry messages
+- No prometheus output for topic-tagged metrics
+
+And the resources we still consume for hibernated topics:
+- Disk space for any segments that had not expired before we hibernated
+- Memory for the topic's frozen metadata
+- Bandwidth to awake and move the data on rebalance or node decommission
+
+## Interaction with other features
+
+Any 'scrub' feature will probably end up waking partitions up to
+scrub them, but that is manageable as long as the scrub mechanism
+properly limits the number of partitions to scrub at once, and groups
+the partitions together by partition when scrubbing them.
+
+Partition movement (especially during node decommission) will require
+waking up the topic.  This should not create a thundering herd during
```

Comment:
definitely waking will be the easiest, but I could see a specialization that didn't require waking. but that's sci-fi stuff right now.